odl/trafos/backends/pywt_bindings.py

```diff
@@ -0,0 +1,523 @@
+        The maximum number of usable scales can be determined
+        by `pywt.dwt_max_level`.
+    pad_mode : string, optional
+        Signal extention mode. Possible extension modes are:
+
+        ``'constant'``: Fill with ``pad_const``.
+
+        ``'symmetric'``: Reflect at the boundaries, not doubling the
+        outmost values.
+
+        ``'periodic'``: Fill in values from the other side, keeping
+        the order.
+
+        ``'order0'``: Extend constantly with the outmost values
+        (ensures continuity).
+
+        ``'order1'``: Extend with constant slope (ensures continuity of
+        the first derivative). This requires at least 2 values along
+        each axis where padding is applied.
+
+        ``'pywt_periodic'``:  like ``'periodic'`` padding but gives
+        the smallest possible number of decomposition coefficients.
+        Only available with ``impl='pywt'``, See `pywt.MODES.modes`.
+
+    Returns
+    -------
+    shape_list : list
+        List containing the shapes of the wavelet (approximation
+        and detail) coefficients at different scaling levels.
+
+        ``shape_list[0]`` = shape of approximation coefficients at
+        the coarsest level,
+
+        ``shape_list[1]`` = shape of the detailed coefficients at
+        the coarsest level,
+
+        ``shape_list[N]`` = shape of the detailed coefficients at
+        the finest level,
+
+        ``shape_list[N+1]`` = shape of the original image,
+
+        ``N`` =  number of scaling levels
+    """
+    if len(shape) not in (1, 2, 3):
+        raise ValueError('shape must have length 1, 2 or 3, got {}'
+                         ''.format(len(shape)))
+
+    wbasis = pywt_wbasis(wbasis)
+    pywt_mode = PAD_MODES_ODL2PYWT[pad_mode]
+
+    max_level = pywt.dwt_max_level(shape[0], filter_len=wbasis.dec_len)
+    if nscales > max_level:
+        raise ValueError('too many scaling levels, got {}, maximum useful '
+                         'level is {}'
+                         ''.format(nscales, max_level))
+
+    # dwt_coeff_len calculates the number of coefficients at the next
+    # scaling level given the input shape, the length of the filter and
+    # the applied mode.
+    # We use this in the following way (per dimension):
+    # - length[0] = original data length
+    # - length[n+1] = dwt_coeff_len(length[n], ...)
+    # - until n = nscales
+    shape_list = [shape]
+    for scale in range(nscales):
+        shp = tuple(pywt.dwt_coeff_len(n, filter_len=wbasis.dec_len,
+                                       mode=pywt_mode)
+                    for n in shape_list[scale])
+        shape_list.append(shp)
+
+    # Add a duplicate of the last entry for the approximation coefficients
+    shape_list.append(shape_list[-1])
+
+    # We created the list in reversed order compared to what pywt expects
+    shape_list.reverse()
+    return shape_list
+
+
+def pywt_coeff_to_array(coeff, shape_list):
```

Comment:
This needs a reference on why the f it is in this order and not some other order. Simply saying "this is how pywt orders them" should be fine.